{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The goal is to annotate MSigDB (https://www.gsea-msigdb.org/gsea/msigdb) and SynGO (https://syngoportal.org/) gene sets to CHR:FROM-TO coordinates (in GRC37 build).\n",
    "From MSigDB I'm interested in the C5 category, excluding HPO sub-category - which leaves Biological Process (BP), Cellular Component (CC) and Molecular Function (MF) ontologies, in total 10402 genesets. SynGO adds additional 73 gene sets (on top of what's already in MSigDB, and excluding small genesets of less than 5 genes). These genesets combined have 19366 unique genes.\n",
    "\n",
    "I'm aware of four different ways of refering to a gene:\n",
    "(1) gene symbol\n",
    "(2) Ensembl ID\n",
    "(3) Entrez ID (from NCBI)\n",
    "(4) HGNC ID\n",
    "\n",
    "C5 category in MSigGB uses Entrez ID, while SynGO uses HGNC IDs. Historically I used to get coordinates from Ensembl, because of convenient bulk export web interface ( http://grch37.ensembl.org/biomart/martview/ ).  So I've been trying to map Entrez IDs and HGNC IDs to Ensembl IDs. \n",
    "\n",
    "The file ``mart_export_14jan2022.txt`` in this repository is downloaded using the link above, after selecting the following columns - which include both HGNC and Entrez IDs. However the mapping is sparse and quite ambiguous.\n",
    "```\n",
    "Gene stable ID,Gene stable ID version,Chromosome/scaffold name,Gene description,Gene name,\n",
    "Gene type,Gene Synonym,Gene start (bp),Gene end (bp),HGNC ID,NCBI gene (formerly Entrezgene) ID\n",
    "```\n",
    "\n",
    "This is easy for SynGO, as it comes with ``syngo_genes.xlsx`` containing a handy table to map HGNC to both Entrez and Ensembl. Unfortunately that table covers only 1233 genes, and I can't use it to map genes in MSigDB ongologies. I also found https://syngoportal.org/convert.html which does a great job of mapping Entrez IDs <-> HGNC IDs (also for all genes in MSigDB). I have the map in ``SynGO_id_convert/idmap.xlsx``. Unfortunately this tool doesn't map to Ensembl ID.\n",
    "\n",
    "My current solution is to use EnsemblID -> EntrezID mapping provided by Ensembl (part of ``mart_Export_14jan2022.txt``). When EntrezID is missing, I tried to put it from ``SynGO_id_convert/idmap.xlsx``, matching based on HGNC ID. This mapped in total 19151 out of 19366 genes, leavning 215 genes unmaped. I haven't tried to use gene symbols to map the remaining genes.\n",
    "\n",
    "I'm somewhat concerned to leave out 215 genes.\n",
    "\n",
    "The conclusion is that EntrezID should be mapped to CHR:FROM-TO coordinates from NCBI database, without involving Ensembl. Sort of obvious, but I didn't expect that mapping humane genes across databases is going to be so difficult :)\n",
    "\n",
    "UPD January 24\n",
    "\n",
    "The solution turned out to be simple - it's quite easy to query NCBI databases, using esearch/efilter/efetch utilities as described here https://www.ncbi.nlm.nih.gov/books/NBK179288/ (super helpful docs!).  Some other alternatives that didn't work: https://www.ncbi.nlm.nih.gov/sites/batchentrez - helpful but can't query more than 1000 at a time, and also can't fetch hg19 coordinates. https://ftp.ncbi.nlm.nih.gov/genomes/all/annotation_releases/9606/105.20201022/ - useful resource but I couldn't figure out whether it contained gene-level information in a suitable format.\n",
    "\n",
    "I query genes for each chromosome as follows:\n",
    "```\n",
    "for i in {1..22}\n",
    "do\n",
    "esearch -db gene -query \"Homo sapiens [ORGN] AND $i [CHR]\" | efilter -status alive -type coding | efetch -format docsum > esearch.chr$i.coding.results;\n",
    "done\n",
    "```\n",
    "Then also for X, Y, and MT chromosomes:\n",
    "```\n",
    "esearch -db gene -query \"Homo sapiens [ORGN] AND X [CHR]\" | efilter -status alive -type coding | efetch -format docsum > esearch.chrX.coding.results;\n",
    "esearch -db gene -query \"Homo sapiens [ORGN] AND Y [CHR]\" | efilter -status alive -type coding | efetch -format docsum > esearch.chrY.coding.results;\n",
    "esearch -db gene -query \"Homo sapiens [ORGN] AND MT [CHR]\" | efilter -status alive -type coding | efetch -format docsum > esearch.chrMT.coding.results;\n",
    "```\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chri 1: 6368 genes in total, 230 without location history\n",
      "chri 2: 4821 genes in total, 160 without location history\n",
      "chri 3: 3779 genes in total, 153 without location history\n",
      "chri 4: 2962 genes in total, 278 without location history\n",
      "chri 5: 3252 genes in total, 343 without location history\n",
      "chri 6: 3858 genes in total, 462 without location history\n",
      "chri 7: 3342 genes in total, 135 without location history\n",
      "chri 8: 2534 genes in total, 98 without location history\n",
      "chri 9: 2660 genes in total, 92 without location history\n",
      "chri 10: 2575 genes in total, 78 without location history\n",
      "chri 11: 3398 genes in total, 132 without location history\n",
      "chri 12: 2910 genes in total, 88 without location history\n",
      "chri 13: 1586 genes in total, 85 without location history\n",
      "chri 14: 2336 genes in total, 86 without location history\n",
      "chri 15: 2158 genes in total, 106 without location history\n",
      "chri 16: 2341 genes in total, 90 without location history\n",
      "remove 2 malformed lines\n",
      "chri 17: 2920 genes in total, 110 without location history\n",
      "chri 18: 1159 genes in total, 49 without location history\n",
      "chri 19: 2905 genes in total, 82 without location history\n",
      "chri 20: 1578 genes in total, 49 without location history\n",
      "chri 21: 911 genes in total, 35 without location history\n",
      "chri 22: 1456 genes in total, 58 without location history\n",
      "chri X: 2615 genes in total, 168 without location history\n",
      "chri Y: 653 genes in total, 45 without location history\n",
      "chri MT: 110 genes in total, 73 without location history\n",
      "save info on 61837 genes to EntrezID.LocationHist.txt\n",
      "chri 1: 2024 genes in total, 7 without location history\n",
      "chri 2: 1234 genes in total, 0 without location history\n",
      "chri 3: 1044 genes in total, 0 without location history\n",
      "chri 4: 745 genes in total, 1 without location history\n",
      "chri 5: 858 genes in total, 1 without location history\n",
      "chri 6: 1027 genes in total, 6 without location history\n",
      "chri 7: 914 genes in total, 0 without location history\n",
      "chri 8: 683 genes in total, 1 without location history\n",
      "chri 9: 764 genes in total, 1 without location history\n",
      "chri 10: 719 genes in total, 0 without location history\n",
      "chri 11: 1292 genes in total, 3 without location history\n",
      "chri 12: 1020 genes in total, 7 without location history\n",
      "chri 13: 326 genes in total, 1 without location history\n",
      "chri 14: 605 genes in total, 2 without location history\n",
      "chri 15: 587 genes in total, 0 without location history\n",
      "chri 16: 847 genes in total, 2 without location history\n",
      "remove 2 malformed lines\n",
      "chri 17: 1159 genes in total, 3 without location history\n",
      "chri 18: 264 genes in total, 0 without location history\n",
      "chri 19: 1417 genes in total, 1 without location history\n",
      "chri 20: 528 genes in total, 0 without location history\n",
      "chri 21: 234 genes in total, 1 without location history\n",
      "chri 22: 448 genes in total, 1 without location history\n",
      "chri X: 854 genes in total, 5 without location history\n",
      "chri Y: 66 genes in total, 2 without location history\n",
      "chri MT: 39 genes in total, 26 without location history\n",
      "save info on 19608 genes to EntrezID.LocationHist.coding.txt\n"
     ]
    }
   ],
   "source": [
    "# read gene information from NCBI\n",
    "# produces files \n",
    "# EntrezID.LocationHist.txt - info on 61837 genes (coding and pseudo)\n",
    "# EntrezID.LocationHist.coding.txt - info on 19608 genes (coding only)\n",
    "# the only filtering at this point is on having location information\n",
    "# here is an example of gene that doesn't have known location info https://www.ncbi.nlm.nih.gov/gene/?term=401191\n",
    "\n",
    "import xmltodict\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "for suffix in ['', '.coding']:\n",
    "    dfs=[]\n",
    "    for chri in list(range(1, 23)) + ['X', 'Y', 'MT']:\n",
    "        num_genes = 0\n",
    "        num_genes_without_location_hist = 0\n",
    "\n",
    "        fname=f'entrez/esearch.chr{chri}{suffix}.results'\n",
    "        xml = open(fname).readlines()\n",
    "        orig_lines = len(xml)\n",
    "        xml = [x for x in xml if (\"'\" not in x)]\n",
    "        diff_lines = orig_lines - len(xml)\n",
    "        if diff_lines>0: print(f'remove {diff_lines} malformed lines')\n",
    "        xml = xmltodict.parse('\\n'.join(['<Root>'] + xml[2:] + ['</Root>']))\n",
    "        DocumentSummarySet = xml['Root']['DocumentSummarySet']\n",
    "        DocumentSummarySet = DocumentSummarySet if isinstance(DocumentSummarySet, list) else [DocumentSummarySet]\n",
    "        for dx in DocumentSummarySet:\n",
    "            for gene in dx['DocumentSummary']:\n",
    "                num_genes += 1\n",
    "                if 'LocationHist' not in gene: \n",
    "                    num_genes_without_location_hist += 1; continue\n",
    "                x=gene['LocationHist']['LocationHistType']\n",
    "                df = pd.DataFrame(x) if isinstance(x, list) else pd.DataFrame(x, index=[0])\n",
    "                df['Id'] = gene['Id']\n",
    "                df['Name'] = gene['Name']\n",
    "                if 'GenomicInfo' in gene:\n",
    "                    x = gene['GenomicInfo']['GenomicInfoType']\n",
    "                    x = x if isinstance(x, list) else [x]\n",
    "                    df['ExonCount'] = x[0]['ExonCount']\n",
    "                else:\n",
    "                    df['ExonCount'] = None\n",
    "                dfs.append(df)\n",
    "        print(f'chri {chri}: {num_genes} genes in total, {num_genes_without_location_hist} without location history')\n",
    "    df=pd.concat(dfs)\n",
    "    df['Id'] = df['Id'].astype(int)\n",
    "    fname = f'EntrezID.LocationHist{suffix}.txt'\n",
    "    df.to_csv(fname,sep='\\t', index=False)\n",
    "    print(f'save info on {len(df[\"Id\"].unique())} genes to {fname}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "reading EntrezID.LocationHist.txt\n",
      "reading EntrezID.LocationHist.coding.txt\n",
      "\n",
      "61837 genes break down into coding and non-coding genes:\n",
      "pseudo    42229\n",
      "coding    19608\n",
      "Name: GeneType, dtype: int64\n",
      "\n",
      "47866 genes after filtering on  AnnotationRelease 105.20201022 and primary assembly\n",
      "(excluding NCBI alternate loci;  see https://www.ncbi.nlm.nih.gov/assembly/model/ for more info)\n",
      "pseudo    28596\n",
      "coding    19270\n",
      "Name: GeneType, dtype: int64\n",
      "\n",
      "45221 genes after filtering onfiltering NCBI genes on chr1..22 (exclude X, Y, MT)\n",
      "pseudo    26844\n",
      "coding    18377\n",
      "Name: GeneType, dtype: int64\n",
      "filter genes that are not present in the reference(s) within 10.0 KB window - this step may take some time...\n",
      "reading /home/oleksanf/github/mixer_private_docker/reference/hrc_EUR_qc/hrc_chr{}_EUR_qc.bim...\n",
      "\t44726 genes remain after excluding 495 genes which do not overlap with /home/oleksanf/github/mixer_private_docker/reference/hrc_EUR_qc/hrc_chr{}_EUR_qc.bim\n",
      "reading /home/oleksanf/github/mixer_private_docker/reference/ukb_EUR_qc/ukb_imp_chr{}_v3_qc.bim...\n",
      "\t44247 genes remain after excluding 479 genes which do not overlap with /home/oleksanf/github/mixer_private_docker/reference/ukb_EUR_qc/ukb_imp_chr{}_v3_qc.bim\n",
      "\n",
      "44247 genes after on UKB and HRC references\n",
      "pseudo    26046\n",
      "coding    18201\n",
      "Name: GeneType, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# Filter Enterez genes\n",
    "# - grc37 annotations release, primary assembly only\n",
    "# - autosomes only (not X, Y, MT)\n",
    "# - with SNPs in UKB and HRC references\n",
    "# Save resulting genes to EntrezID.QCed.txt\n",
    "\n",
    "import xmltodict\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "do_filter_by_reference = True\n",
    "window_size = 10000\n",
    "\n",
    "fname = 'EntrezID.LocationHist.txt'\n",
    "print(f'reading {fname}')\n",
    "ncbi_gene = pd.read_csv(fname, sep='\\t', dtype=str)\n",
    "\n",
    "fname = 'EntrezID.LocationHist.coding.txt'\n",
    "print(f'reading {fname}')\n",
    "ncbi_gene_coding = pd.read_csv(fname, sep='\\t', dtype=str)\n",
    "\n",
    "ncbi_gene_coding['GeneType'] = 'coding'\n",
    "ncbi_gene = pd.merge(ncbi_gene, ncbi_gene_coding[['Id', 'GeneType']].drop_duplicates(), on='Id', how='left')\n",
    "ncbi_gene.loc[ncbi_gene['GeneType'].isnull(), 'GeneType'] = 'pseudo'\n",
    "\n",
    "print(f'\\n{len(ncbi_gene[\"Id\"].unique())} genes break down into coding and non-coding genes:')\n",
    "print(ncbi_gene[['Id', 'GeneType']].drop_duplicates()['GeneType'].value_counts())\n",
    "\n",
    "ncbi_gene['Id']=ncbi_gene['Id'].astype(int)\n",
    "ncbi_gene['ChrStart']=ncbi_gene['ChrStart'].astype(int)\n",
    "ncbi_gene['ChrStop']=ncbi_gene['ChrStop'].astype(int)\n",
    "\n",
    "idx = (ncbi_gene['AnnotationRelease']=='105.20201022') & ncbi_gene['ChrAccVer'].str.startswith('NC_')\n",
    "ncbi_gene = ncbi_gene[idx].copy()\n",
    "print(f'\\n{len(ncbi_gene[\"Id\"].unique())} genes after filtering on  AnnotationRelease 105.20201022 and primary assembly\\n(excluding NCBI alternate loci;  see https://www.ncbi.nlm.nih.gov/assembly/model/ for more info)')\n",
    "print(ncbi_gene[['Id', 'GeneType']].drop_duplicates()['GeneType'].value_counts())\n",
    "\n",
    "ncbi_gene = ncbi_gene[[x.split('.')[0] not in ['NC_000023', 'NC_000024', 'NC_012920'] for x in ncbi_gene['ChrAccVer'].values]].copy()\n",
    "print(f'\\n{len(ncbi_gene[\"Id\"].unique())} genes after filtering onfiltering NCBI genes on chr1..22 (exclude X, Y, MT)')\n",
    "print(ncbi_gene[['Id', 'GeneType']].drop_duplicates()['GeneType'].value_counts())\n",
    "\n",
    "ncbi_gene['CHR'] = [int(x.replace('NC_', '').split('.')[0]) for x in ncbi_gene['ChrAccVer'].values]\n",
    "ncbi_gene['FROM'] = np.minimum(ncbi_gene['ChrStart'].values, ncbi_gene['ChrStop'].values)\n",
    "ncbi_gene['TO'] = np.maximum(ncbi_gene['ChrStart'].values, ncbi_gene['ChrStop'].values)\n",
    "ncbi_gene['EZID'] =  ncbi_gene['Id']\n",
    "if do_filter_by_reference:\n",
    "    df = ncbi_gene.copy()\n",
    "    # now filter genes that aren't present in the reference\n",
    "    print(f'filter genes that are not present in the reference(s) within {window_size/1000} KB window - this step may take some time...')\n",
    "    for ref_file in ['/home/oleksanf/github/mixer_private_docker/reference/hrc_EUR_qc/hrc_chr{}_EUR_qc.bim', '/home/oleksanf/github/mixer_private_docker/reference/ukb_EUR_qc/ukb_imp_chr{}_v3_qc.bim']:\n",
    "        print(f'reading {ref_file}...')\n",
    "        df['SNPS_in_ref'] = 0\n",
    "        for chri in range(1, 23):\n",
    "            #print(chri)\n",
    "            ref=pd.read_csv(ref_file.format(chri), sep='\\t', header=None, names='CHR SNP GP BP A1 A2'.split())\n",
    "            for idx in df[df['CHR']==chri].index:\n",
    "                a=df.loc[idx, 'FROM']\n",
    "                b=df.loc[idx, 'TO']\n",
    "                df.loc[idx, 'SNPS_in_ref'] = np.sum((ref['BP']>=a-window_size) & (ref['BP'] <=b+window_size))\n",
    "        print(f'\\t{np.sum(df.SNPS_in_ref!=0)} genes remain after excluding {np.sum(df.SNPS_in_ref==0)} genes which do not overlap with {ref_file}')\n",
    "        df = df[df['SNPS_in_ref'] != 0].copy()\n",
    "        del df['SNPS_in_ref']\n",
    "    ncbi_gene = df \n",
    "print(f'\\n{len(ncbi_gene[\"Id\"].unique())} genes after on UKB and HRC references')\n",
    "print(ncbi_gene[['Id', 'GeneType']].drop_duplicates()['GeneType'].value_counts())\n",
    "\n",
    "ncbi_gene.rename(columns={'Name':'GENE'}, inplace=True)\n",
    "ncbi_gene[['EZID', 'GENE', 'GeneType', 'CHR', 'FROM', 'TO']].to_csv('EntrezID.QCed.txt', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "reading msigdb_v7.5.xml...\n",
      "\ttotal genesets in MSIGDB/GENESET: 33365\n",
      "\tafter filtering on C5: 15473\n",
      "\tafter excluding SUB_CATEGORY_CODE HPO: 10402\n",
      "reading SynGO/syngo_ontologies.xls...\n",
      "\t293 genesets in total\n",
      "\t184 after filtering out genesets found in msigdb\n",
      "\t73 after filtering genesets with less than 5 genes (same threshold as applied in MSigDB)\n",
      "\t10475 genesets in total after integrating 73 syngo genesets with msigdb\n",
      "\t19366 unique genes across genesets\n",
      "reading EntrezID.QCed.txt...\n",
      "\t44247 genes in NCBI (after QC)\n",
      "\t18201 genes in NCBI (after QC, conding genes only)\n",
      "after merging gene and geneset definitions:\n",
      "\tgenesets:     10472\n",
      "\tgenes:        16787\n",
      "select 6350 gene sets for leave-one-gene-out analysis; 10.84 genes per gene set on average\n"
     ]
    }
   ],
   "source": [
    "import xmltodict\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "do_save = True\n",
    "\n",
    "import os\n",
    "if not os.path.exists('msigdb_v7.5.xml'):\n",
    "    os.system('gunzip -c msigdb_v7.5.xml.gz >msigdb_v7.5.xml')\n",
    "\n",
    "# read gene set definitions\n",
    "# documentation: https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/MSigDB_v7.0_Release_Notes\n",
    "#fname = \"msigdb_v7.0.xml\"\n",
    "fname = \"msigdb_v7.5.xml\"\n",
    "print(f'reading {fname}...')\n",
    "xml = xmltodict.parse(open(fname).read())\n",
    "cols = 'STANDARD_NAME EXACT_SOURCE MEMBERS MEMBERS_EZID CATEGORY_CODE SUB_CATEGORY_CODE'\n",
    "msigdb = pd.DataFrame([tuple([x['@'+col] for col in cols.split()]) for x in xml['MSIGDB']['GENESET']], columns=cols.split())\n",
    "print(f'\\ttotal genesets in MSIGDB/GENESET: {len(msigdb)}')\n",
    "\n",
    "msigdb=msigdb[msigdb['CATEGORY_CODE'].isin(['C5'])]  # ['C2', 'C5'])]\n",
    "print(f'\\tafter filtering on C5: {len(msigdb)}')\n",
    "\n",
    "msigdb=msigdb[~msigdb['SUB_CATEGORY_CODE'].isin(['HPO'])] \n",
    "print(f'\\tafter excluding SUB_CATEGORY_CODE HPO: {len(msigdb)}')\n",
    "\n",
    "assert len(msigdb['STANDARD_NAME'].unique())== len(msigdb) # check that all standard names are unique\n",
    "genesets = msigdb[['EXACT_SOURCE', 'STANDARD_NAME', 'MEMBERS_EZID']].copy()\n",
    "\n",
    "fname='SynGO/syngo_ontologies.xls'\n",
    "print(f'reading {fname}...')\n",
    "syngo_onto=pd.read_excel(fname)\n",
    "syngo_gene=pd.read_excel('SynGO/syngo_genes.xls')\n",
    "syngo_anno=pd.read_excel('SynGO/syngo_annotations.xls')\n",
    "\n",
    "# convert from hgnc gene ids to EZID\n",
    "hgnc_to_enterez = dict((hgnc_id, entrez_id) for hgnc_id, entrez_id in syngo_gene[['hgnc_id', 'entrez_id']].values)\n",
    "syngo_onto['MEMBERS_EZID'] = [\n",
    "    ','.join([str(hgnc_to_enterez[hgnc]) for hgnc in genes_hgnc_id.split(';')])\n",
    "    for genes_hgnc_id in syngo_onto['genes - hgnc_id'].values]\n",
    "\n",
    "syngo_onto = syngo_onto[['GO term ID', 'GO term name', 'MEMBERS_EZID']]\n",
    "syngo_onto.columns=['EXACT_SOURCE', 'STANDARD_NAME', 'MEMBERS_EZID']\n",
    "syngo_onto['STANDARD_NAME'] = ['SYNGO_' + '_'.join(x.upper().split()) for x in syngo_onto['STANDARD_NAME'].values]\n",
    "syngo_onto['GENE_COUNT'] = [len(x.split(',')) for x in syngo_onto['MEMBERS_EZID']]\n",
    "\n",
    "print(f'\\t{len(syngo_onto)} genesets in total')\n",
    "\n",
    "syngo_onto=syngo_onto[~syngo_onto['EXACT_SOURCE'].isin(genesets['EXACT_SOURCE'].values)].copy()\n",
    "print(f'\\t{len(syngo_onto)} after filtering out genesets found in msigdb')\n",
    "\n",
    "syngo_onto  = syngo_onto[syngo_onto['GENE_COUNT'] >= 5].copy()\n",
    "del syngo_onto['GENE_COUNT']\n",
    "print(f'\\t{len(syngo_onto)} after filtering genesets with less than 5 genes (same threshold as applied in MSigDB)')\n",
    "\n",
    "genesets = pd.concat([genesets, syngo_onto])\n",
    "print(f'\\t{len(genesets)} genesets in total after integrating {len(syngo_onto)} syngo genesets with msigdb')\n",
    "\n",
    "if do_save:\n",
    "    genesets.to_csv('gsa-mixer-geneset-with-exact-source_27jan2022.csv', sep='\\t', index=False)\n",
    "\n",
    "# convert to long format\n",
    "go_id = 'STANDARD_NAME'\n",
    "genesets = pd.DataFrame(genesets['MEMBERS_EZID'].str.split(',').tolist(), index=genesets[go_id]).stack()\n",
    "genesets = genesets.reset_index([0, go_id])\n",
    "genesets.columns = [go_id, 'EZID']\n",
    "genesets['EZID']=genesets['EZID'].astype(int)\n",
    "genesets.rename(columns={go_id:'GO'}, inplace=True)\n",
    "print(f'\\t{len(genesets[\"EZID\"].unique())} unique genes across genesets')\n",
    "\n",
    "fname = 'EntrezID.QCed.txt'\n",
    "print(f'reading {fname}...')\n",
    "ncbi_gene=pd.read_csv(fname, sep='\\t')\n",
    "print(f'\\t{len(ncbi_gene)} genes in NCBI (after QC)')\n",
    "\n",
    "ncbi_gene = ncbi_gene[ncbi_gene['GeneType']=='coding'].copy()\n",
    "print(f'\\t{len(ncbi_gene)} genes in NCBI (after QC, conding genes only)')\n",
    "\n",
    "print('after merging gene and geneset definitions:')\n",
    "genesets = pd.merge(genesets, ncbi_gene, on='EZID', how='inner')\n",
    "print('\\tgenesets:     {}'.format(len(genesets['GO'].unique())))\n",
    "print('\\tgenes:        {}'.format(len(genesets['GENE'].unique())))\n",
    "\n",
    "# generate all_genes, coding_genes and pseudo_genes categories (for the baseline model)\n",
    "df2=ncbi_gene.loc[ncbi_gene['GeneType']=='coding', ['GENE', 'CHR', 'FROM', 'TO']].copy()\n",
    "df2['GO'] = 'coding_genes'\n",
    "\n",
    "# put each gene as it's own gene set ('solo')\n",
    "df4=ncbi_gene.copy()\n",
    "df4['GO'] = df4['GENE']\n",
    "\n",
    "# gene sets (GO + SynGO)\n",
    "df5=genesets[['GO', 'GENE', 'CHR', 'FROM', 'TO']].copy()\n",
    "\n",
    "# gene sets (GO + SynG0) leave-one-out analysis for gene sets with up to 25 genes\n",
    "assert len([x for x in list(set(df5['GO'].values)) if '_excl_' in x]) == 0 # check we can use '_excl_' as a separator between geneset and 'loo' gene being excluded\n",
    "df5_count = df5[['GO', 'GENE']].groupby(['GO'], as_index=False).count()\n",
    "df5_count = df5_count[df5_count['GENE']<=25]\n",
    "geneset_for_loo_analysis = df5_count['GO'].values\n",
    "print(f'select {len(geneset_for_loo_analysis)} gene sets for leave-one-gene-out analysis; {df5_count[\"GENE\"].mean():.2f} genes per gene set on average')\n",
    "df5loo = df5[df5['GO'].isin(geneset_for_loo_analysis)].copy()\n",
    "df5loo = pd.merge(df5loo[['GO', 'GENE']].rename(columns={'GENE':'excl_GENE'}), df5loo)\n",
    "df5loo = df5loo[df5loo['excl_GENE'] != df5loo['GENE']]\n",
    "df5loo['GO'] = [f'{go}_excl_{gene}' for go, gene in zip(df5loo['GO'].values, df5loo['excl_GENE'].values) ]\n",
    "del df5loo['excl_GENE']\n",
    "\n",
    "if do_save:\n",
    "    # save the results (GSA MiXeR format)\n",
    "    pd.concat([df2], sort=True)[['GO', 'GENE', 'CHR', 'FROM', 'TO']].to_csv('gsa-mixer-baseline-annot_10mar2023.csv', sep='\\t', index=False)\n",
    "    pd.concat([df2, df4], sort=True)[['GO', 'GENE', 'CHR', 'FROM', 'TO']].to_csv('gsa-mixer-gene-annot_10mar2023.csv', sep='\\t', index=False)\n",
    "    pd.concat([df2, df5], sort=True)[['GO', 'GENE', 'CHR', 'FROM', 'TO']].to_csv('gsa-mixer-geneset-annot_10mar2023.csv', sep='\\t', index=False)\n",
    "    pd.concat([df2, df5, df5loo], sort=True)[['GO', 'GENE', 'CHR', 'FROM', 'TO']].to_csv('gsa-mixer-genesetLOO-annot_10mar2023.csv', sep='\\t', index=False)      \n",
    "    pd.concat([df2, df4, df5], sort=True)[['GO', 'GENE', 'CHR', 'FROM', 'TO']].to_csv('gsa-mixer-hybrid-annot_10mar2023.csv', sep='\\t', index=False)\n",
    "    pd.concat([df2, df4, df5, df5loo], sort=True)[['GO', 'GENE', 'CHR', 'FROM', 'TO']].to_csv('gsa-mixer-hybridLOO-annot_10mar2023.csv', sep='\\t', index=False)      \n",
    "\n",
    "    # save the resuls (MAGMA format)\n",
    "    ncbi_gene['STRAND'] = '+'\n",
    "    ncbi_gene[['GENE', 'CHR', 'FROM', 'TO', 'STRAND', 'GENE']].to_csv('magma-gene-annot_10mar2023.csv', sep='\\t', index=False)\n",
    "    genesets[['GO', 'GENE']].groupby(['GO'], as_index=False).agg({'GENE': ' '.join}).to_csv('magma-geneset-annot_10mar2023.csv', sep='\\t', index=False, header=None)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
